src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -243,3 +243,195 @@ def predict(
             for c in self.components
         ]
         return sum([w * pmf for w, pmf in zip(self.ws, scaled_component_pmfs)])
+
+
+class DiracMixtureSinglePublisherModel(ReachCurve):
+    def __init__(
+        self,
+        data: List[ReachPoint],
+        universe_reach_ratio: float = 3,
+        ncomponents: int = 200,
+    ):
+        """Constructs a Dirac mixture single publisher model.
+
+        Args:
+            data:  A list consisting of the single ReachPoint to which the model
+                is to be fit.
+            universe_reach_ratio:  We start with the universe size that is this
+                much times the 1+ reach for fitting the model.  By "start with"
+                it means the universe size may be modified if the specified
+                universe size does not give a good fit.  See the comments in the
+                _fit method for details.
+            ncomponents:  Number of components in the Poisson mixture.  Based on
+                our experience, 100 components are enough, and more components
+                are also fine (they will not introduce overfitting).
+        """
+        if len(data) != 1:
+            raise ValueError("Exactly one ReachPoint must be specified")
+        if data[0].impressions[0] < 0.001:
+            raise ValueError("Attempt to create model with 0 impressions")
+        if data[0].impressions[0] < data[0].reach(1):
+            raise ValueError(
+                "Cannot have a model with fewer impressions than reached people"
+            )
+        self._data = data
+        self._reach_point = data[0]
+        if data[0].spends:
+            self._cpi = data[0].spends[0] / data[0].impressions[0]
+        else:
+            self._cpi = None
+        self.N = data[0].reach(1) * universe_reach_ratio
```

Comment:
I think the universe size should be a property of the ReachCurve!

Reply:
Thank you Matt.

For simplicity, I'm using the following code design:
- Add `universe_size` as an optional input to `DataSet`.  If it's not given, then set DataSet.universe_size as 2 * DataSet.maximum_reach.   This is not principled, but makes life easier.
- `HaloSimulator` reads a `DataSet` and generate different `ReachPoint`.  Every `ReachPoint` generated here now carry the `universe_size` from `HaloSimulator._data_set`.
- `DiracMixtureSinglePublisherModel` takes a `ReachPoint` as input.  Raise error if this `ReachPoint` does not have a `universe_size`.

PLMK WYT!

Reply:
@matthewclegg @jiayu-google 
Yeah, I guessed that probably the original choice of no universe size was made so that we can have the possibility of fitting to unknown universe sizes.

However, I don't really see a good chance of getting reasonable estimates on universe size using reach curves. But that is fine for now, let's leave this question open and discuss later.